website/blog/2018-09-10-account-recovery-in-firefox-accounts.md

```diff
@@ -0,0 +1,147 @@
+a user’s kB in an encrypted bundle on our servers. The recovery
+key is part of the encryption key that allows the user to decrypt
+the bundle that is stored. It is generated on the user’s web
+client and will only be shown once. FxA never sends the recovery
+key and it is expected that the user prints or stores this key
+in a safe location.
+
+During a password reset, the recovery key is hashed into a recovery
+key ID. This ID can be used to retrieve the encrypted bundle and
+decrypt it to get the original kB. Finally, when the password is
+reset, the original kB and new password are used to  generate
+the updated encryption keys.
+
+The security requirements for the feature:
+
+1. The recovery key must be short enough that it can be reliably stored and retrieved by the user without corruption on a printout.
+2. The recovery key must be long enough that brute-force guessing a token is infeasible (online or offline)
+3. The combination of the recovery key plus some information held by the FxA server must be sufficient to recover kB.
+4. It must be infeasible to recover kB from the recovery key alone, without retrieving additional data from the Firefox Account server. (This is a safeguard against leakage of the token from wherever the user stores it).
+5. It must be infeasible to recover kB from data held by the Firefox Account server alone, without also possessing the recovery key.  (This is a safeguard against Mozilla, and leaks of our server-side database state).
+6. It must be possible for the holder of the recovery key to convince the Firefox Account server that they hold a valid key, without revealing the key itself. (This is a safeguard Mozilla, and leaks of our server-side database state.)
+
+### Deriving keys and JSON Web Key
+
+A core component to account recovery is how we generate the
+different kinds of keys used during encryption. Below is a
+summary of those keys and how they are created.
+
+* Recovery Key (recoveryKey)
+  * 16 digit [Crockford Base32](https://en.wikipedia.org/wiki/Base32) that is generated client-side and shown to the user only once
+* Recovery Key ID (recoveryKeyId)
+  * recoveryKeyId = HKDF(recoveryKey, salt=uid, info=“fxa recovery fingerprint”, length=16)
+  * Opaque key generated from the recovery key and sent to our servers
+* Encryption Key (encryptionKey)
+  * encryptionKey = HKDF(recoveryKey, salt=uid, info=“fxa recovery encrypt key”, length=32)
+  * Key used to encrypt kB
+* Unique user ID (uid)
+  * User’s unique account identifier
+
+Our servers will ever only know and interact with the recovery
+key ID. By doing this, FxA safeguards itself from ever knowing
+the raw recovery key and therefore would not be able to derive
+kB to decrypt any data. Even if there was a database breach,
+an attacker would still not know the original recovery key.
+
+Using the values above, a JSON Web Key [(JWK)](https://en.wikipedia.org/wiki/JSON_Web_Token) is created with the following properties.
```

Comment:
Do you want to link to a JWT article here? The [JWK spec](https://tools.ietf.org/html/rfc7517) is a bit dry but pretty approachable as far as specs go, and the [introduction](https://tools.ietf.org/html/rfc7517#section-1) has most of the applicable info.